pkg/logutil/log_util.go

```diff
@@ -0,0 +1,32 @@
+// Copyright 2018 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package logutil
+
+// LogVerbosity is the level of verbosity of a particular log statement.
+// Usage: log.V(LogVerbosity).Info(...)
+type LogVerbosity = int
+
+// The default log verbosity, given to the most important log statements,
+// is 0: log.V(0).Info(...) == log.Info(...)
+//
+// HACK: zapr doesn't calculate log levels correctly, so we must set log
+// level integers incorrectly (according to logr) for now so they map to zapr's.
+const (
+	// VDebug == -1. log.V(VDebug).Info(...) == log.Debug(...)
+	VDebug LogVerbosity = iota - 1
+	_
+	// VWarn == 1. log.V(VWarn).Info(...) == log.Warn(...)
+	VWarn
```

Comment:
Going on a bit of exposition here:

So logr levels have no concept of Warn, Fatal or anything else. It only has Info and Error.
Logr's Info logs [decrease in importance](https://github.com/kubernetes-sigs/controller-runtime/blob/53fc44b56078cd095b11bd44cfa0288ee4cf718f/vendor/github.com/go-logr/logr/logr.go#L70-L73) as the levels(0,1,2,...) go up e.g 0 is considered standard Info, 1 could be considered Debug(less important), 2 is like more Debug(even less important). 
No levels map to Warn or Fatal which are generally more important than Info and Debug.

Zap levels are the inverse. -2(More Debug) -1(Debug), 0 (Info), 1(Warning) ...
https://godoc.org/go.uber.org/zap/zapcore#Level

But since zapr just flips the logr level it will only end up using the following zap levels: 0(Info), -1(Debug), -2(More Debug) ...
https://github.com/go-logr/zapr/blob/master/zapr.go#L133-L134
https://github.com/go-logr/zapr/blob/master/zapr.go#L34-L37

So zapr's mapping is correct.

In practice because the runtime's helper uses either [Development](https://godoc.org/go.uber.org/zap#NewDevelopment) or [Production](https://godoc.org/go.uber.org/zap#NewProduction) configs for zapr, we can only see 2 (zap)levels of logs.
- Production => zap.Info(0) and higher
  - so only logr.V(0) => zap.V(0). Remember logr.V(-1) is illegal.
- Development => zap.Debug(-1) and higher
  - so that gives us logr.V(1) => zap.V(-1) and  logr.V(0) => zap.V(0) and 
https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/runtime/log/log.go#L39-L48

Since logr.V(2) or zap.V(-2) is a custom Debug level those logs will not show up. So logr.V(2) is pointless if we're using the runtime's [`ZapLogger(development bool)`](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/runtime/log/log.go#L36) method.

That's why the runtime only has 2 levels of logs statements right now. 
[Info or logr.V(0)](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/internal/controller/controller.go#L121) and [Debug logr.V(1)](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/internal/controller/controller.go#L232)

Reply:
> And if we must have consts then they should only be VInfo=0 and VDebug=1.
Warnings or VWarn is going against the logr since there is no level for warnings and V(1) is not a warning.

We agreed on warning to be 0 in the meeting as summed up
https://github.com/operator-framework/operator-sdk/issues/503#issuecomment-432764973

But yes, reading the logr docs again and:

> I restrict the API to just 2 types of logs: info and error.

So should we then just go with info and error only, as it also has no concept of debug either and not assign "names" to numbers as suggested by logr docs?

Reply:
> I would prefer not to use these consts and just use logr.V(0)( which is justlogr.Info()) and logr.V(1).

👍 

I think it is just easier to deal with IMO.